### ISSUE TITLE
Bug: Only typeof in proc body does not account for closure var

### ISSUE THREAD

**Issue by straight-shoota** [crystal-lang/crystal#6040]
```cr
foo = "foo"
proc = -> do
  typeof(foo) # => String
  # foo # (uncomment for correct result)
end
foo = nil
proc.call
```

`typeof(foo)` should be `String | Nil` here. This is correctly reported if `foo` is directly referenced in the proc body. But it should also work if it is only used in a `typeof`.

**Comment by asterite** [comment]
Yes, `typeof` is just for compile-time, so `foo` isn't really closured here, as its value isn't used at runtime. So not a bug.

**Comment by straight-shoota** [comment]
Hm, technically yes. But wouldn't it make sense to treat `foo` as being closured? It is used in the closure, even if it is only referenced in the `typeof`. That's what I'd expect here. It's not super important but would be consistent and the least surprise.

**Comment by ysbaddaden** [comment]
We miss a `help-wanted` label on issues.

**Comment by straight-shoota** [comment]
@ysbaddaden There is [community:to-implement](https://github.com/crystal-lang/crystal/labels/community%3Ato-implement).

**Comment by ysbaddaden** [comment]
That's... not really explicit.
